### PR TITLE
Add :on_duplicate_key_update support for Postgres

### LINKED ISSUE #187: Support INSERT ... ON CONFLICT DO NOTHING/UPDATE (also known as UPSERT) for PostgreSQL
The feature exists already for MySQL databases and can be used with the :on_duplicate_key_update parameter. Upcoming versions of PostgreSQL will feature a similar functionality: 

http://git.postgresql.org/gitweb/?p=postgresql.git;a=commit;h=168d5805e4c08bed7b95d351bf097cff7c07dd65

It would be cool to see this feature in activerecord-import, also for PostgreSQL databases.

### PR DESCRIPTION
**Pull request by jkowens**
Postgres 9.5 added UPSERT support. Resolves #187.

Provides the following syntax:

```
Country.import [country], validate: false, on_duplicate_key_update: { conflict_target: :country, columns: [:updated_at] }
```

`:conflict_target` takes either a single value or an array of column names that make up the unique constraint and `:columns` could be either an array or hash similar to the existing MySQL functionality.

### CODE REVIEW

**Comment by zdennis** [comment]
OMG @jkowens, I :heart: you right now.

**Comment by jkowens** [comment]
Haha! Let me know if you find any issues or have any suggestions for this PR, and I'll do what I can. Thanks :beers:

**Comment by triptec** [comment]
Any progress here?

**Comment by jkowens** [comment]
I've rebased master and made some tweaks to the PR. If anyone else has a chance to review and test that would probably help @zdennis out a bit :wink:

File: lib/activerecord-import/adapters/postgresql_adapter.rb [hunk 1]
```diff
@@ -30,6 +31,67 @@ def post_sql_statements( table_name, options ) # :nodoc:
...
+    end
+  end
+
+  # Returns a generated ON CONFLICT DO UPDATE statement given the passed
+  # in +options+.
+  def sql_for_on_duplicate_key_update( table_name, options={} ) # :nodoc:
+    return unless options.is_a?( Hash )
+
+    key = Array( options[:key] ).join( ', ' )
+    raise ArgumentError, 'Attribute :key not found' if key.empty?
```
Comment (zdennis) [inline lib/activerecord-import/adapters/postgresql_adapter.rb]: It looks like `:key` represents the [ON CONFLICT](http://www.postgresql.org/docs/9.5/static/sql-insert.html#SQL-ON-CONFLICT)  `conflict_target`. If that's the case can we make this optional?

File: lib/activerecord-import/import.rb [hunk 2]
```diff
@@ -225,13 +225,31 @@ def support_setting_primary_key_of_imported_objects?
...
+    #
+    # The :on_duplicate_key_update option is a hash with :key and
+    # :columns attributes. Below is an example:
+    #
+    #   BlogPost.import columns, values, :on_duplicate_key_update=>{ :key => :id, :columns => [ :date_modified, :content, :author ] }
+    #
+    # ==== :key
+    #
+    # The :key attribute specifies the columns that make up the unique constraint and can be
+    # a single column or an array of column names.
```
Comment (zdennis) [inline lib/activerecord-import/import.rb]: WDYT about specifically mentioning what key maps to in PostgreSQL?

File: lib/activerecord-import/import.rb [hunk 3]
```diff
@@ -225,13 +225,31 @@ def support_setting_primary_key_of_imported_objects?
...
     # model. Below is an example:
     #
     #   BlogPost.import columns, attributes, :on_duplicate_key_update=>{ :title => :title }
     #
+    # == On Duplicate Key Update (Postgres 9.5+)
+    #
+    # The :on_duplicate_key_update option is a hash with :key and
+    # :columns attributes. Below is an example:
+    #
+    #   BlogPost.import columns, values, :on_duplicate_key_update=>{ :key => :id, :columns => [ :date_modified, :content, :author ] }
```
Comment (zdennis) [inline lib/activerecord-import/import.rb]: I wonder if we should provide additional example(s) for `:key` usage. The ones that come to mind are:

* Default: no `key` specified. (If this ends up being applicable)
* Passing `key` to use existing index
* Passing `key` to use existing column(s) as index? (I think this is what you have above)

There may be other options such as passing in PG-compatible `index_expression`(s) but I'm not entirely sure if that will already work or if it'd be more work (which I don't think should block this PR getting in).

File: test/support/postgresql/import_examples.rb [hunk 4]
```diff
@@ -127,3 +127,174 @@ def should_support_postgresql_import_functionality
...
+
+        it "should raise an ArgumentError" do
+          assert_raise ArgumentError do
+            perform_import
+          end
+        end
+      end
+    end
+  end
+end
```
Comment (zdennis) [inline test/support/postgresql/import_examples.rb]: @jkowens, do you think we can make the default behavior and interface for PostgreSQL compatible with MySQL? If so, I'd love to move all of the baseline examples into a single spec and then leave any examples that are MySQL or PostgreSQL specific (incompatible APIs) into their respective specs. 

Thoughts?

**Comment by zdennis** [comment]
@jkowens, this is great stuff. Great work!  

My main hope before we merge this in and release is it to make the baseline APIs b/w MySQL and PostgreSQL compatible or to share the knowledge of why they cannot be so we can document that reasoning.

I will also work on getting TravisCi to run with PostgreSQL 9.5 so we can have this running by default.

File: test/postgresql/import_test.rb [hunk 5]
```diff
@@ -1,4 +1,7 @@
 require File.expand_path(File.dirname(__FILE__) + '/../test_helper')
 require File.expand_path(File.dirname(__FILE__) + '/../support/postgresql/import_examples')
 
+pg_version = ENV["PG_VERSION"] || "0"
```
Comment (zdennis) [inline test/postgresql/import_test.rb]: Reminder to self that I should probably create a test readme file so options like this can be surfaced. 

Out of curiosity what did you use to determine to how to set up and run the test environment? Just figure it out by looking through code or was there any information that was helpful any where in the repos or on the internets?

File: lib/activerecord-import/adapters/postgresql_adapter.rb (hunk 1 above)
Comment (jkowens) [inline lib/activerecord-import/adapters/postgresql_adapter.rb]: Yes, that's correct. The ON CONFLICT clause requires a conflict_target, unless the conflict_action is DO NOTHING. Currently the only time that conflict_action would be DO NOTHING is if :columns is an empty array *and* the table does not have update timestamp columns that automatically get assigned. I can update the code to make :key optional in that scenario

Unfortunately, as far as I can tell, this requirement makes matching the MySQL interface not possible :frowning:

File: lib/activerecord-import/import.rb (hunk 2 above)
Comment (jkowens) [inline lib/activerecord-import/import.rb]: Sure thing. Do you think it would be better to rename :key :conflict_target?

File: test/postgresql/import_test.rb (hunk 5 above)
Comment (jkowens) [inline test/postgresql/import_test.rb]: I had to look through the code a bit, but this [wiki entry](https://github.com/zdennis/activerecord-import/wiki/How-to-run-the-tests) really helped.

File: lib/activerecord-import/adapters/postgresql_adapter.rb (hunk 1 above)
Comment (zdennis) [inline lib/activerecord-import/adapters/postgresql_adapter.rb]: Ahh ok. WDYT about defaulting key to :id?

Comment (jkowens) [inline lib/activerecord-import/adapters/postgresql_adapter.rb]: Hm I'm wondering if that could lead to more confusion? Maybe someone assumes if it's the primary key they don't need to specify the conflict_target, but it turns out they prefix the id field with table name. Or they expect it to work just like MySQL. I suppose good documentation could remedy that though.

Comment (zdennis) [inline lib/activerecord-import/adapters/postgresql_adapter.rb]: What about using AR::Base#primary_key ?

File: lib/activerecord-import/import.rb (hunk 2 above)
Comment (zdennis) [inline lib/activerecord-import/import.rb]: :thumbsup: I like that.

File: lib/activerecord-import/adapters/postgresql_adapter.rb (hunk 1 above)
Comment (jkowens) [inline lib/activerecord-import/adapters/postgresql_adapter.rb]: Looks like that should work. I'll make that change so key has a default value (if table has a primary key).

**Comment by jkowens** [comment]
@zdennis I still need to get to documenting options and examples. I will try to do that soon.

File: lib/activerecord-import/import.rb (hunk 3 above)
Comment (jkowens) [inline lib/activerecord-import/import.rb]: Could you explain what you mean by passing `key` to use existing index? Postgres has the ability to determine what index maps to a list column names. The Postgres docs say that a constraint name can be explicitly specified, but is not recommended. Should I look into supporting that option?

File: test/support/postgresql/import_examples.rb (hunk 4 above)
Comment (jkowens) [inline test/support/postgresql/import_examples.rb]: @zdennis using your suggestion to default `:conflict_target` to the primary key, it is now possible to provide an array to :on_duplicate_key_update for Postgres.

`:on_duplicate_key_update => [:column1, :column2]`

I've made the array option a baseline example and pulled that out into a separate shared examples file. The hash option is included in both of the adapter specific examples.

File: lib/activerecord-import/import.rb (hunk 3 above)
Comment (jkowens) [inline lib/activerecord-import/import.rb]: @zdennis I've added support for explicitly naming the conflicting constraint and I've added an example of that to the comments.

**Comment by jkowens** [comment]
@zdennis hopefully I'm getting closer with these latest updates. Let me know if you have any further suggestions and I will try to get them implemented. I can squash all these commits down into one when the time comes.

File: lib/activerecord-import/adapters/postgresql_adapter.rb [hunk 6]
```diff
@@ -30,6 +31,92 @@ def post_sql_statements( table_name, options ) # :nodoc:
...
+    return unless arg.is_a?( Hash )
+
+    constraint_name = arg[ :constraint_name ]
+    conflict_target = arg[ :conflict_target ]
+    columns         = arg.fetch( :columns, [] )
+
+    sql = " ON CONFLICT "
+    conflict_target = sql_for_conflict_target( conflict_target, constraint_name )
+
+    if columns.respond_to?( :empty? ) && columns.empty?
```
Comment (triptec) [inline lib/activerecord-import/adapters/postgresql_adapter.rb]: Is there any test for this block? I've been trying to do something like this:
Category.import categories, :on_duplicate_key_update=>{ :conflict_target => [:name], :columns => [], validate: false }
So if the category exists I don't want to update it, just DO NOTHING

Comment (jkowens) [inline lib/activerecord-import/adapters/postgresql_adapter.rb]: Yes, but the code currently looks to see if the table has updated_on or updated_at timestamp columns and if it does those are automatically added to the list of columns, so it's unlikely `columns` will empty. I'm thinking this code was put in place for MySQL since its on duplicate key update clause does not have a "do nothing" option, so columns were required. That brings up a good question, since Postgres does have a "DO NOTHING" option, should we not automatically add the update columns? @zdennis what are your thoughts?

Comment (zdennis) [inline lib/activerecord-import/adapters/postgresql_adapter.rb]: @jkowens said:
> Yes, but the code currently looks to see if the table has updated_on or updated_at timestamp columns and if it does those are automatically added to the list of columns, so it's unlikely columns will be empty.

While likely true it's not guaranteed so I agree with @triptec that we should have a test for this.

> I'm thinking this code was put in place for MySQL since its on duplicate key update clause does not have a "do nothing" option, so columns were required

MySQL has INSERT IGNORE which allows I believe achieve the same end-result: ignore inserting rows that fail a required constraint check. This is supported in activerecord-import by passing the `ignore: true` option to `import(...)`. 

In hindsight, this is a terribly named option in activerecord-import.

> That brings up a good question, since Postgres does have a "DO NOTHING" option, should we not automatically add the update columns?

I think we add them if the caller passed in other columns, but only then. See below `on_duplicate_key_ignore` suggestion. 

@triptec said:
> I've been trying to do something like this:
Category.import categories, :on_duplicate_key_update=>{ :conflict_target => [:name], :columns => [], validate: false }

@triptec, so you don't really want to update anything, you just want to ignore duplicate records, yes? If so, what about something like this (which doesn't exist yet):

```
Category.import categories, :on_duplicate_key_ignore=>{ :conflict_target => [:name] }, validate: false
```

This would make it clear that you intend to ignore duplicates and make help us cleanly separate this from the case where we automatically update timestamps when present.

Comment (triptec) [inline lib/activerecord-import/adapters/postgresql_adapter.rb]: That would be great.

Comment (jkowens) [inline lib/activerecord-import/adapters/postgresql_adapter.rb]: @zdennis @triptec I took a stab at adding `:on_duplicate_key_ignore` for Postgres. As Zach mentioned, MySQL does offer the `ignore: true` option. Passing a hash with a :conflict_target as shown above works, but I thought just passing `true` would more than likely be what most used for this scenario:

```
Category.import categories, on_duplicate_key_ignore: true, validate: false
```

`ON CONFLICT DO NOTHING` is the one time Postgres lets us get away without specifying a conflict target.

File: test/postgresql/import_test.rb (hunk 5 above)
Comment (jkowens) [inline test/postgresql/import_test.rb]: @zdennis I refactored this so that ENV variable PG_VERSION is no longer required. I didn't realize I could grab PostgreSQL version from `ActiveRecord::ConnectionAdapters::PostgreSQLAdapter#postresql_version`.

File: lib/activerecord-import/adapters/postgresql_adapter.rb (hunk 6 above)
Comment (jkowens) [inline lib/activerecord-import/adapters/postgresql_adapter.rb]: I think `:on_duplicate_key_ignore` could potentially be an alias for  MySQL `:ignore`, but I think that probably deserves its own PR.

File: test/support/shared_examples/on_duplicate_key_update.rb [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+        end
+
+        context "using symbol column names" do
+          let(:update_columns){ [ :title, :author_email_address, :parent_id ] }
+          should_support_on_duplicate_key_update
+          should_update_fields_mentioned
+        end
+      end
+
+      context "given array of model instances with :validation checks turned off" do
```
Comment (jkowens) [inline test/support/shared_examples/on_duplicate_key_update.rb]: @zdennis is there a need to duplicate tests for importing an array of models to validate :on_duplicate_key_update functionality? If not, I was going to skinny these examples down a bit.

File: test/support/postgresql/import_examples.rb (hunk 4 above)
Comment (zdennis) [inline test/support/postgresql/import_examples.rb]: :thumbsup:

File: test/support/shared_examples/on_duplicate_key_update.rb (hunk 7 above)
Comment (zdennis) [inline test/support/shared_examples/on_duplicate_key_update.rb]: I don't know off the top of my head so I'd say go for it. :)

File: lib/activerecord-import/adapters/postgresql_adapter.rb [hunk 8]
```diff
@@ -24,12 +27,113 @@ def next_value_for_sequence(sequence_name)
...
+  end
+
+  # Returns a generated ON CONFLICT DO NOTHING statement given the passed
+  # in +args+.
+  def sql_for_on_duplicate_key_ignore( table_name, *args ) # :nodoc:
+    arg = args.first
+    if arg.is_a?( Hash )
+      conflict_target = sql_for_conflict_target( arg ) if arg.is_a?( Hash )
+    end
+    " ON CONFLICT #{conflict_target}DO NOTHING"
```
Comment (zdennis) [inline lib/activerecord-import/adapters/postgresql_adapter.rb]: Does this potentially need a space before `DO NOTHING`?

File: lib/activerecord-import/import.rb [hunk 9]
```diff
@@ -165,15 +165,19 @@ def support_setting_primary_key_of_imported_objects?
...
-    # * +on_duplicate_key_update+ - an Array or Hash, tells import to \
-    #    use MySQL's ON DUPLICATE KEY UPDATE ability. See On Duplicate\
-    #    Key Update below.
+    # * +ignore+ - true|false, tells import to use MySQL's INSERT IGNORE
+    #    to discard records that contain duplicate keys.
+    # * +on_duplicate_key_ignore+ - true|false, tells import to use
+    #    Postgres 9.5+ ON CONFLICT DO NOTHING.
+    # * +on_duplicate_key_update+ - an Array or Hash, tells import to
+    #    use MySQL's ON DUPLICATE KEY UPDATE or Postgres 9.5+ ON CONFLICT
+    #    DO UPDATE ability. See On Duplicate Key Update below.
```
Comment (zdennis) [inline lib/activerecord-import/import.rb]: TY for updating the docs. I think github ate one of your comments, but I agree with you that `on_duplicate_key_ignore` could become an alias for `ignore`, but it should be done in a separate PR.

File: lib/activerecord-import/import.rb [hunk 10]
```diff
@@ -225,13 +229,73 @@ def support_setting_primary_key_of_imported_objects?
...
+    #
+    #   BlogPost.import columns, values, :on_duplicate_key_update=>{ :conflict_target => :slug, :columns => [ :date_modified, :content, :author ] }
+    #
+    # ========  Using a Hash
+    #
+    # The :columns option can be a hash of column names to model attribute name
+    # mappings. This gives you finer grained control over what fields are updated
+    # with what attributes on your model. Below is an example:
+    #
+    #   BlogPost.import columns, attributes, :on_duplicate_key_update=>{ :conflict_target => :slug, :columns => { :title => :title } }
```
Comment (zdennis) [inline lib/activerecord-import/import.rb]: Awesome docs above.

File: lib/activerecord-import/adapters/postgresql_adapter.rb (hunk 8 above)
Comment (jkowens) [inline lib/activerecord-import/adapters/postgresql_adapter.rb]: `sql_for_conflict_target` returns a string with a blank space at the end, so I think it will be ok.

**Comment by zdennis** [comment]
Thanks for all your work on this @jkowens! This is really awesome stuff and I know there are a lot of people who are going to be excited for it. The test failure has to do with `em-synchrony` as it doesn't support ActiveRecord 5.0.0 atm so don't worry about that. 

Is there anything else you wanted to do before I merge this in?

**Comment by jkowens** [comment]
Thanks @zdennis :smile: 

If you are good with what we have here, I can squash the commits into one for you to merge.

**Comment by zdennis** [comment]
Let's do it @jkowens

**Comment by jkowens** [comment]
All set :+1:

**Comment by rwilliams** [comment]
Awesome!

**Comment by zdennis** [comment]
Thank you for your work @jkowens ! I will push out a new gem this weekend and PostgreSQL databases everywhere will be filled with the joy of speedier imports from applications using ActiveRecord.

**Comment by jkowens** [comment]
Yes!! :rocket:

**Comment by fenec** [comment]
:fireworks:

**Comment by zdennis** [comment]
This has been included in the [0.12.0](https://rubygems.org/gems/activerecord-import/versions/0.12.0) release.
